### ISSUE TITLE
Support for OptIn annotaion in kotlin codegen

### ISSUE THREAD

**Issue by Wasabi375** [square/moshi#1135]
I want to use moshi to generate adapters for properties that use a custom `@Experimental` annotation.
Currently this doesn't work because the `toJson` function generated by moshi won't compile as it's missing the `OptIn(Experimental::class)` annotation. 

My suggestion is that moshi could add a `CodegenOptIn` annotation that adds the required `OptIn` to the `toJson` and `fromJson` functions.

**Comment by JakeWharton** [comment]
I don't really understand why we want a new annotation for this. Seems more straightforward to just mirror the experimental annotation out onto the generated code.

**Comment by Wasabi375** [comment]
Yeah, you are totally right. Don't know why thought a separate annotation was necessary.

**Comment by tomridder** [comment]
> I want to use moshi to generate adapters for properties that use a custom `@Experimental` annotation. Currently this doesn't work because the `toJson` function generated by moshi won't compile as it's missing the `OptIn(Experimental::class)` annotation.
> 
> My suggestion is that moshi could add a `CodegenOptIn` annotation that adds the required `OptIn` to the `toJson` and `fromJson` functions.
> 
> Also if you like the idea I can look into implementing this if you want. Let me know.

What are the situations where @Experimental annotation needs to be added in the code generated by using @JsonClass?
since moshi doesn't support inline class yet

**Comment by ZacSweers** [comment]
PR's welcome on this one 👍

**Comment by Ayman-Alkurdi** [comment]
Hi @ZacSweers 
Could you please review my PR that addresses this issue? Thanks!
